// thirdweb/dist/esm/extensions/erc4337/__generated__/IAccountFactory/read/totalAccounts.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x58451f97";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `totalAccounts` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `totalAccounts` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isTotalAccountsSupported } from "thirdweb/extensions/erc4337";
 *
 * const supported = await isTotalAccountsSupported(contract);
 * ```
 */
export async function isTotalAccountsSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the totalAccounts function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC4337
 * @example
 * ```ts
 * import { decodeTotalAccountsResult } from "thirdweb/extensions/erc4337";
 * const result = decodeTotalAccountsResult("...");
 * ```
 */
export function decodeTotalAccountsResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "totalAccounts" function on the contract.
 * @param options - The options for the totalAccounts function.
 * @returns The parsed result of the function call.
 * @extension ERC4337
 * @example
 * ```ts
 * import { totalAccounts } from "thirdweb/extensions/erc4337";
 *
 * const result = await totalAccounts({
 *  contract,
 * });
 *
 * ```
 */
export async function totalAccounts(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
