// thirdweb/dist/cjs/extensions/erc721/__generated__/DropERC721/write/updateBatchBaseURI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateBatchBaseURI = exports.encodeUpdateBatchBaseURI = exports.encodeUpdateBatchBaseURIParams = exports.isUpdateBatchBaseURISupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xde903ddd";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "_index",
    },
    {
        type: "string",
        name: "_uri",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `updateBatchBaseURI` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `updateBatchBaseURI` method is supported.
 * @extension ERC721
 * @example
 * ```ts
 * import { isUpdateBatchBaseURISupported } from "thirdweb/extensions/erc721";
 *
 * const supported = await isUpdateBatchBaseURISupported(contract);
 * ```
 */
async function isUpdateBatchBaseURISupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isUpdateBatchBaseURISupported = isUpdateBatchBaseURISupported;
/**
 * Encodes the parameters for the "updateBatchBaseURI" function.
 * @param options - The options for the updateBatchBaseURI function.
 * @returns The encoded ABI parameters.
 * @extension ERC721
 * @example
 * ```ts
 * import { encodeUpdateBatchBaseURIParams } "thirdweb/extensions/erc721";
 * const result = encodeUpdateBatchBaseURIParams({
 *  index: ...,
 *  uri: ...,
 * });
 * ```
 */
function encodeUpdateBatchBaseURIParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.index, options.uri]);
}
exports.encodeUpdateBatchBaseURIParams = encodeUpdateBatchBaseURIParams;
/**
 * Encodes the "updateBatchBaseURI" function into a Hex string with its parameters.
 * @param options - The options for the updateBatchBaseURI function.
 * @returns The encoded hexadecimal string.
 * @extension ERC721
 * @example
 * ```ts
 * import { encodeUpdateBatchBaseURI } "thirdweb/extensions/erc721";
 * const result = encodeUpdateBatchBaseURI({
 *  index: ...,
 *  uri: ...,
 * });
 * ```
 */
function encodeUpdateBatchBaseURI(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeUpdateBatchBaseURIParams(options).slice(2));
}
exports.encodeUpdateBatchBaseURI = encodeUpdateBatchBaseURI;
/**
 * Prepares a transaction to call the "updateBatchBaseURI" function on the contract.
 * @param options - The options for the "updateBatchBaseURI" function.
 * @returns A prepared transaction object.
 * @extension ERC721
 * @example
 * ```ts
 * import { updateBatchBaseURI } from "thirdweb/extensions/erc721";
 *
 * const transaction = updateBatchBaseURI({
 *  contract,
 *  index: ...,
 *  uri: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function updateBatchBaseURI(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.index, resolvedOptions.uri];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.updateBatchBaseURI = updateBatchBaseURI;
