// thirdweb/dist/cjs/extensions/modular/__generated__/ModularCore/write/cancelOwnershipHandover.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cancelOwnershipHandover = exports.isCancelOwnershipHandoverSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x54d1f13d";
const FN_INPUTS = [];
const FN_OUTPUTS = [];
/**
 * Checks if the `cancelOwnershipHandover` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `cancelOwnershipHandover` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isCancelOwnershipHandoverSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isCancelOwnershipHandoverSupported(contract);
 * ```
 */
async function isCancelOwnershipHandoverSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isCancelOwnershipHandoverSupported = isCancelOwnershipHandoverSupported;
/**
 * Prepares a transaction to call the "cancelOwnershipHandover" function on the contract.
 * @param options - The options for the "cancelOwnershipHandover" function.
 * @returns A prepared transaction object.
 * @extension MODULAR
 * @example
 * ```ts
 * import { cancelOwnershipHandover } from "thirdweb/extensions/modular";
 *
 * const transaction = cancelOwnershipHandover();
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function cancelOwnershipHandover(options) {
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.cancelOwnershipHandover = cancelOwnershipHandover;
