// thirdweb/dist/cjs/extensions/modular/__generated__/ModularCore/write/completeOwnershipHandover.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.completeOwnershipHandover = exports.encodeCompleteOwnershipHandover = exports.encodeCompleteOwnershipHandoverParams = exports.isCompleteOwnershipHandoverSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xf04e283e";
const FN_INPUTS = [
    {
        name: "pendingOwner",
        type: "address",
        internalType: "address",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `completeOwnershipHandover` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `completeOwnershipHandover` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isCompleteOwnershipHandoverSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isCompleteOwnershipHandoverSupported(contract);
 * ```
 */
async function isCompleteOwnershipHandoverSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isCompleteOwnershipHandoverSupported = isCompleteOwnershipHandoverSupported;
/**
 * Encodes the parameters for the "completeOwnershipHandover" function.
 * @param options - The options for the completeOwnershipHandover function.
 * @returns The encoded ABI parameters.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeCompleteOwnershipHandoverParams } "thirdweb/extensions/modular";
 * const result = encodeCompleteOwnershipHandoverParams({
 *  pendingOwner: ...,
 * });
 * ```
 */
function encodeCompleteOwnershipHandoverParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.pendingOwner]);
}
exports.encodeCompleteOwnershipHandoverParams = encodeCompleteOwnershipHandoverParams;
/**
 * Encodes the "completeOwnershipHandover" function into a Hex string with its parameters.
 * @param options - The options for the completeOwnershipHandover function.
 * @returns The encoded hexadecimal string.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeCompleteOwnershipHandover } "thirdweb/extensions/modular";
 * const result = encodeCompleteOwnershipHandover({
 *  pendingOwner: ...,
 * });
 * ```
 */
function encodeCompleteOwnershipHandover(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeCompleteOwnershipHandoverParams(options).slice(2));
}
exports.encodeCompleteOwnershipHandover = encodeCompleteOwnershipHandover;
/**
 * Prepares a transaction to call the "completeOwnershipHandover" function on the contract.
 * @param options - The options for the "completeOwnershipHandover" function.
 * @returns A prepared transaction object.
 * @extension MODULAR
 * @example
 * ```ts
 * import { completeOwnershipHandover } from "thirdweb/extensions/modular";
 *
 * const transaction = completeOwnershipHandover({
 *  contract,
 *  pendingOwner: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function completeOwnershipHandover(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.pendingOwner];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.completeOwnershipHandover = completeOwnershipHandover;
