// thirdweb/dist/cjs/extensions/modular/__generated__/ModularCore/write/renounceOwnership.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.renounceOwnership = exports.isRenounceOwnershipSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x715018a6";
const FN_INPUTS = [];
const FN_OUTPUTS = [];
/**
 * Checks if the `renounceOwnership` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `renounceOwnership` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isRenounceOwnershipSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isRenounceOwnershipSupported(contract);
 * ```
 */
async function isRenounceOwnershipSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isRenounceOwnershipSupported = isRenounceOwnershipSupported;
/**
 * Prepares a transaction to call the "renounceOwnership" function on the contract.
 * @param options - The options for the "renounceOwnership" function.
 * @returns A prepared transaction object.
 * @extension MODULAR
 * @example
 * ```ts
 * import { renounceOwnership } from "thirdweb/extensions/modular";
 *
 * const transaction = renounceOwnership();
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function renounceOwnership(options) {
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.renounceOwnership = renounceOwnership;
