// thirdweb/dist/cjs/extensions/modular/__generated__/ModularCore/write/renounceRoles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.renounceRoles = exports.encodeRenounceRoles = exports.encodeRenounceRolesParams = exports.isRenounceRolesSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x183a4f6e";
const FN_INPUTS = [
    {
        name: "roles",
        type: "uint256",
        internalType: "uint256",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `renounceRoles` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `renounceRoles` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isRenounceRolesSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isRenounceRolesSupported(contract);
 * ```
 */
async function isRenounceRolesSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isRenounceRolesSupported = isRenounceRolesSupported;
/**
 * Encodes the parameters for the "renounceRoles" function.
 * @param options - The options for the renounceRoles function.
 * @returns The encoded ABI parameters.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeRenounceRolesParams } "thirdweb/extensions/modular";
 * const result = encodeRenounceRolesParams({
 *  roles: ...,
 * });
 * ```
 */
function encodeRenounceRolesParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.roles]);
}
exports.encodeRenounceRolesParams = encodeRenounceRolesParams;
/**
 * Encodes the "renounceRoles" function into a Hex string with its parameters.
 * @param options - The options for the renounceRoles function.
 * @returns The encoded hexadecimal string.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeRenounceRoles } "thirdweb/extensions/modular";
 * const result = encodeRenounceRoles({
 *  roles: ...,
 * });
 * ```
 */
function encodeRenounceRoles(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeRenounceRolesParams(options).slice(2));
}
exports.encodeRenounceRoles = encodeRenounceRoles;
/**
 * Prepares a transaction to call the "renounceRoles" function on the contract.
 * @param options - The options for the "renounceRoles" function.
 * @returns A prepared transaction object.
 * @extension MODULAR
 * @example
 * ```ts
 * import { renounceRoles } from "thirdweb/extensions/modular";
 *
 * const transaction = renounceRoles({
 *  contract,
 *  roles: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function renounceRoles(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.roles];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.renounceRoles = renounceRoles;
