// thirdweb/dist/cjs/extensions/modular/__generated__/ModularCore/write/requestOwnershipHandover.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.requestOwnershipHandover = exports.isRequestOwnershipHandoverSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x25692962";
const FN_INPUTS = [];
const FN_OUTPUTS = [];
/**
 * Checks if the `requestOwnershipHandover` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `requestOwnershipHandover` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isRequestOwnershipHandoverSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isRequestOwnershipHandoverSupported(contract);
 * ```
 */
async function isRequestOwnershipHandoverSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isRequestOwnershipHandoverSupported = isRequestOwnershipHandoverSupported;
/**
 * Prepares a transaction to call the "requestOwnershipHandover" function on the contract.
 * @param options - The options for the "requestOwnershipHandover" function.
 * @returns A prepared transaction object.
 * @extension MODULAR
 * @example
 * ```ts
 * import { requestOwnershipHandover } from "thirdweb/extensions/modular";
 *
 * const transaction = requestOwnershipHandover();
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function requestOwnershipHandover(options) {
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.requestOwnershipHandover = requestOwnershipHandover;
