// thirdweb/dist/cjs/extensions/modular/__generated__/ModularCore/write/revokeRoles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.revokeRoles = exports.encodeRevokeRoles = exports.encodeRevokeRolesParams = exports.isRevokeRolesSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x4a4ee7b1";
const FN_INPUTS = [
    {
        name: "user",
        type: "address",
        internalType: "address",
    },
    {
        name: "roles",
        type: "uint256",
        internalType: "uint256",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `revokeRoles` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `revokeRoles` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isRevokeRolesSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isRevokeRolesSupported(contract);
 * ```
 */
async function isRevokeRolesSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isRevokeRolesSupported = isRevokeRolesSupported;
/**
 * Encodes the parameters for the "revokeRoles" function.
 * @param options - The options for the revokeRoles function.
 * @returns The encoded ABI parameters.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeRevokeRolesParams } "thirdweb/extensions/modular";
 * const result = encodeRevokeRolesParams({
 *  user: ...,
 *  roles: ...,
 * });
 * ```
 */
function encodeRevokeRolesParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.user, options.roles]);
}
exports.encodeRevokeRolesParams = encodeRevokeRolesParams;
/**
 * Encodes the "revokeRoles" function into a Hex string with its parameters.
 * @param options - The options for the revokeRoles function.
 * @returns The encoded hexadecimal string.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeRevokeRoles } "thirdweb/extensions/modular";
 * const result = encodeRevokeRoles({
 *  user: ...,
 *  roles: ...,
 * });
 * ```
 */
function encodeRevokeRoles(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeRevokeRolesParams(options).slice(2));
}
exports.encodeRevokeRoles = encodeRevokeRoles;
/**
 * Prepares a transaction to call the "revokeRoles" function on the contract.
 * @param options - The options for the "revokeRoles" function.
 * @returns A prepared transaction object.
 * @extension MODULAR
 * @example
 * ```ts
 * import { revokeRoles } from "thirdweb/extensions/modular";
 *
 * const transaction = revokeRoles({
 *  contract,
 *  user: ...,
 *  roles: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function revokeRoles(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.user, resolvedOptions.roles];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.revokeRoles = revokeRoles;
