// thirdweb/dist/cjs/extensions/modular/__generated__/ModularCore/write/transferOwnership.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transferOwnership = exports.encodeTransferOwnership = exports.encodeTransferOwnershipParams = exports.isTransferOwnershipSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xf2fde38b";
const FN_INPUTS = [
    {
        name: "newOwner",
        type: "address",
        internalType: "address",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `transferOwnership` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `transferOwnership` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isTransferOwnershipSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isTransferOwnershipSupported(contract);
 * ```
 */
async function isTransferOwnershipSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isTransferOwnershipSupported = isTransferOwnershipSupported;
/**
 * Encodes the parameters for the "transferOwnership" function.
 * @param options - The options for the transferOwnership function.
 * @returns The encoded ABI parameters.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeTransferOwnershipParams } "thirdweb/extensions/modular";
 * const result = encodeTransferOwnershipParams({
 *  newOwner: ...,
 * });
 * ```
 */
function encodeTransferOwnershipParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.newOwner]);
}
exports.encodeTransferOwnershipParams = encodeTransferOwnershipParams;
/**
 * Encodes the "transferOwnership" function into a Hex string with its parameters.
 * @param options - The options for the transferOwnership function.
 * @returns The encoded hexadecimal string.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeTransferOwnership } "thirdweb/extensions/modular";
 * const result = encodeTransferOwnership({
 *  newOwner: ...,
 * });
 * ```
 */
function encodeTransferOwnership(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeTransferOwnershipParams(options).slice(2));
}
exports.encodeTransferOwnership = encodeTransferOwnership;
/**
 * Prepares a transaction to call the "transferOwnership" function on the contract.
 * @param options - The options for the "transferOwnership" function.
 * @returns A prepared transaction object.
 * @extension MODULAR
 * @example
 * ```ts
 * import { transferOwnership } from "thirdweb/extensions/modular";
 *
 * const transaction = transferOwnership({
 *  contract,
 *  newOwner: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function transferOwnership(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.newOwner];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.transferOwnership = transferOwnership;
