// thirdweb/dist/esm/extensions/modular/__generated__/ModularCore/write/cancelOwnershipHandover.js
import { prepareContractCall } from "../../../../../transaction/prepare-contract-call.js";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x54d1f13d";
const FN_INPUTS = [];
const FN_OUTPUTS = [];
/**
 * Checks if the `cancelOwnershipHandover` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `cancelOwnershipHandover` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isCancelOwnershipHandoverSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isCancelOwnershipHandoverSupported(contract);
 * ```
 */
export async function isCancelOwnershipHandoverSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Prepares a transaction to call the "cancelOwnershipHandover" function on the contract.
 * @param options - The options for the "cancelOwnershipHandover" function.
 * @returns A prepared transaction object.
 * @extension MODULAR
 * @example
 * ```ts
 * import { cancelOwnershipHandover } from "thirdweb/extensions/modular";
 *
 * const transaction = cancelOwnershipHandover();
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
export function cancelOwnershipHandover(options) {
    return prepareContractCall({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
