// thirdweb/dist/esm/extensions/modular/__generated__/ModularCore/write/installExtension.js
import { prepareContractCall } from "../../../../../transaction/prepare-contract-call.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { once } from "../../../../../utils/promise/once.js";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xaca696f5";
const FN_INPUTS = [
    {
        name: "_extension",
        type: "address",
        internalType: "address",
    },
    {
        name: "_data",
        type: "bytes",
        internalType: "bytes",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `installExtension` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `installExtension` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isInstallExtensionSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isInstallExtensionSupported(contract);
 * ```
 */
export async function isInstallExtensionSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "installExtension" function.
 * @param options - The options for the installExtension function.
 * @returns The encoded ABI parameters.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeInstallExtensionParams } "thirdweb/extensions/modular";
 * const result = encodeInstallExtensionParams({
 *  extension: ...,
 *  data: ...,
 * });
 * ```
 */
export function encodeInstallExtensionParams(options) {
    return encodeAbiParameters(FN_INPUTS, [options.extension, options.data]);
}
/**
 * Encodes the "installExtension" function into a Hex string with its parameters.
 * @param options - The options for the installExtension function.
 * @returns The encoded hexadecimal string.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeInstallExtension } "thirdweb/extensions/modular";
 * const result = encodeInstallExtension({
 *  extension: ...,
 *  data: ...,
 * });
 * ```
 */
export function encodeInstallExtension(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeInstallExtensionParams(options).slice(2));
}
/**
 * Prepares a transaction to call the "installExtension" function on the contract.
 * @param options - The options for the "installExtension" function.
 * @returns A prepared transaction object.
 * @extension MODULAR
 * @example
 * ```ts
 * import { installExtension } from "thirdweb/extensions/modular";
 *
 * const transaction = installExtension({
 *  contract,
 *  extension: ...,
 *  data: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
export function installExtension(options) {
    const asyncOptions = once(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return prepareContractCall({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.extension, resolvedOptions.data];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
