// thirdweb/dist/esm/extensions/modular/__generated__/ModularCore/write/renounceOwnership.js
import { prepareContractCall } from "../../../../../transaction/prepare-contract-call.js";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x715018a6";
const FN_INPUTS = [];
const FN_OUTPUTS = [];
/**
 * Checks if the `renounceOwnership` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `renounceOwnership` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isRenounceOwnershipSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isRenounceOwnershipSupported(contract);
 * ```
 */
export async function isRenounceOwnershipSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Prepares a transaction to call the "renounceOwnership" function on the contract.
 * @param options - The options for the "renounceOwnership" function.
 * @returns A prepared transaction object.
 * @extension MODULAR
 * @example
 * ```ts
 * import { renounceOwnership } from "thirdweb/extensions/modular";
 *
 * const transaction = renounceOwnership();
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
export function renounceOwnership(options) {
    return prepareContractCall({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
