// thirdweb/dist/cjs/extensions/modular/__generated__/IModularExtension/read/getExtensionConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExtensionConfig = exports.decodeGetExtensionConfigResult = exports.isGetExtensionConfigSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x7c173ecc";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        name: "",
        type: "tuple",
        internalType: "struct IExtensionConfig.ExtensionConfig",
        components: [
            {
                name: "registerInstallationCallback",
                type: "bool",
                internalType: "bool",
            },
            {
                name: "requiredInterfaces",
                type: "bytes4[]",
                internalType: "bytes4[]",
            },
            {
                name: "supportedInterfaces",
                type: "bytes4[]",
                internalType: "bytes4[]",
            },
            {
                name: "callbackFunctions",
                type: "tuple[]",
                internalType: "struct IExtensionConfig.CallbackFunction[]",
                components: [
                    {
                        name: "selector",
                        type: "bytes4",
                        internalType: "bytes4",
                    },
                ],
            },
            {
                name: "fallbackFunctions",
                type: "tuple[]",
                internalType: "struct IExtensionConfig.FallbackFunction[]",
                components: [
                    {
                        name: "selector",
                        type: "bytes4",
                        internalType: "bytes4",
                    },
                    {
                        name: "permissionBits",
                        type: "uint256",
                        internalType: "uint256",
                    },
                ],
            },
        ],
    },
];
/**
 * Checks if the `getExtensionConfig` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getExtensionConfig` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isGetExtensionConfigSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isGetExtensionConfigSupported(contract);
 * ```
 */
async function isGetExtensionConfigSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetExtensionConfigSupported = isGetExtensionConfigSupported;
/**
 * Decodes the result of the getExtensionConfig function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULAR
 * @example
 * ```ts
 * import { decodeGetExtensionConfigResult } from "thirdweb/extensions/modular";
 * const result = decodeGetExtensionConfigResult("...");
 * ```
 */
function decodeGetExtensionConfigResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetExtensionConfigResult = decodeGetExtensionConfigResult;
/**
 * Calls the "getExtensionConfig" function on the contract.
 * @param options - The options for the getExtensionConfig function.
 * @returns The parsed result of the function call.
 * @extension MODULAR
 * @example
 * ```ts
 * import { getExtensionConfig } from "thirdweb/extensions/modular";
 *
 * const result = await getExtensionConfig({
 *  contract,
 * });
 *
 * ```
 */
async function getExtensionConfig(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.getExtensionConfig = getExtensionConfig;
