// thirdweb/dist/esm/extensions/modular/__generated__/ModularCore/read/getSupportedCallbackFunctions.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xf147db8a";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        name: "",
        type: "tuple[]",
        internalType: "struct IModularCore.SupportedCallbackFunction[]",
        components: [
            {
                name: "selector",
                type: "bytes4",
                internalType: "bytes4",
            },
            {
                name: "mode",
                type: "uint8",
                internalType: "enum IModularCore.CallbackMode",
            },
        ],
    },
];
/**
 * Checks if the `getSupportedCallbackFunctions` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getSupportedCallbackFunctions` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isGetSupportedCallbackFunctionsSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isGetSupportedCallbackFunctionsSupported(contract);
 * ```
 */
export async function isGetSupportedCallbackFunctionsSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getSupportedCallbackFunctions function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULAR
 * @example
 * ```ts
 * import { decodeGetSupportedCallbackFunctionsResult } from "thirdweb/extensions/modular";
 * const result = decodeGetSupportedCallbackFunctionsResult("...");
 * ```
 */
export function decodeGetSupportedCallbackFunctionsResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getSupportedCallbackFunctions" function on the contract.
 * @param options - The options for the getSupportedCallbackFunctions function.
 * @returns The parsed result of the function call.
 * @extension MODULAR
 * @example
 * ```ts
 * import { getSupportedCallbackFunctions } from "thirdweb/extensions/modular";
 *
 * const result = await getSupportedCallbackFunctions({
 *  contract,
 * });
 *
 * ```
 */
export async function getSupportedCallbackFunctions(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
