// thirdweb/dist/cjs/extensions/erc4337/__generated__/IAccountFactory/read/totalAccounts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.totalAccounts = exports.decodeTotalAccountsResult = exports.isTotalAccountsSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x58451f97";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `totalAccounts` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `totalAccounts` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isTotalAccountsSupported } from "thirdweb/extensions/erc4337";
 *
 * const supported = await isTotalAccountsSupported(contract);
 * ```
 */
async function isTotalAccountsSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isTotalAccountsSupported = isTotalAccountsSupported;
/**
 * Decodes the result of the totalAccounts function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC4337
 * @example
 * ```ts
 * import { decodeTotalAccountsResult } from "thirdweb/extensions/erc4337";
 * const result = decodeTotalAccountsResult("...");
 * ```
 */
function decodeTotalAccountsResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeTotalAccountsResult = decodeTotalAccountsResult;
/**
 * Calls the "totalAccounts" function on the contract.
 * @param options - The options for the totalAccounts function.
 * @returns The parsed result of the function call.
 * @extension ERC4337
 * @example
 * ```ts
 * import { totalAccounts } from "thirdweb/extensions/erc4337";
 *
 * const result = await totalAccounts({
 *  contract,
 * });
 *
 * ```
 */
async function totalAccounts(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.totalAccounts = totalAccounts;
