// thirdweb/dist/cjs/extensions/modular/__generated__/ModularCore/read/ownershipHandoverExpiresAt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ownershipHandoverExpiresAt = exports.decodeOwnershipHandoverExpiresAtResult = exports.encodeOwnershipHandoverExpiresAt = exports.encodeOwnershipHandoverExpiresAtParams = exports.isOwnershipHandoverExpiresAtSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xfee81cf4";
const FN_INPUTS = [
    {
        name: "pendingOwner",
        type: "address",
        internalType: "address",
    },
];
const FN_OUTPUTS = [
    {
        name: "result",
        type: "uint256",
        internalType: "uint256",
    },
];
/**
 * Checks if the `ownershipHandoverExpiresAt` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `ownershipHandoverExpiresAt` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isOwnershipHandoverExpiresAtSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isOwnershipHandoverExpiresAtSupported(contract);
 * ```
 */
async function isOwnershipHandoverExpiresAtSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isOwnershipHandoverExpiresAtSupported = isOwnershipHandoverExpiresAtSupported;
/**
 * Encodes the parameters for the "ownershipHandoverExpiresAt" function.
 * @param options - The options for the ownershipHandoverExpiresAt function.
 * @returns The encoded ABI parameters.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeOwnershipHandoverExpiresAtParams } "thirdweb/extensions/modular";
 * const result = encodeOwnershipHandoverExpiresAtParams({
 *  pendingOwner: ...,
 * });
 * ```
 */
function encodeOwnershipHandoverExpiresAtParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.pendingOwner]);
}
exports.encodeOwnershipHandoverExpiresAtParams = encodeOwnershipHandoverExpiresAtParams;
/**
 * Encodes the "ownershipHandoverExpiresAt" function into a Hex string with its parameters.
 * @param options - The options for the ownershipHandoverExpiresAt function.
 * @returns The encoded hexadecimal string.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeOwnershipHandoverExpiresAt } "thirdweb/extensions/modular";
 * const result = encodeOwnershipHandoverExpiresAt({
 *  pendingOwner: ...,
 * });
 * ```
 */
function encodeOwnershipHandoverExpiresAt(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeOwnershipHandoverExpiresAtParams(options).slice(2));
}
exports.encodeOwnershipHandoverExpiresAt = encodeOwnershipHandoverExpiresAt;
/**
 * Decodes the result of the ownershipHandoverExpiresAt function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULAR
 * @example
 * ```ts
 * import { decodeOwnershipHandoverExpiresAtResult } from "thirdweb/extensions/modular";
 * const result = decodeOwnershipHandoverExpiresAtResult("...");
 * ```
 */
function decodeOwnershipHandoverExpiresAtResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeOwnershipHandoverExpiresAtResult = decodeOwnershipHandoverExpiresAtResult;
/**
 * Calls the "ownershipHandoverExpiresAt" function on the contract.
 * @param options - The options for the ownershipHandoverExpiresAt function.
 * @returns The parsed result of the function call.
 * @extension MODULAR
 * @example
 * ```ts
 * import { ownershipHandoverExpiresAt } from "thirdweb/extensions/modular";
 *
 * const result = await ownershipHandoverExpiresAt({
 *  contract,
 *  pendingOwner: ...,
 * });
 *
 * ```
 */
async function ownershipHandoverExpiresAt(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.pendingOwner],
    });
}
exports.ownershipHandoverExpiresAt = ownershipHandoverExpiresAt;
