// thirdweb/dist/cjs/react/native/ui/components/text.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThemedText = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_native_1 = require("react-native");
function ThemedText({ style, theme, type = "default", ...rest }) {
    return ((0, jsx_runtime_1.jsx)(react_native_1.Text, { style: [
            { color: theme.colors.primaryText },
            type === "default" ? styles.default : undefined,
            type === "title" ? styles.title : undefined,
            type === "defaultSemiBold" ? styles.defaultSemiBold : undefined,
            type === "subtitle" ? styles.subtitle : undefined,
            type === "link" ? styles.link : undefined,
            style,
        ], ...rest }));
}
exports.ThemedText = ThemedText;
const styles = react_native_1.StyleSheet.create({
    default: {
        fontSize: 16,
        lineHeight: 24,
    },
    defaultSemiBold: {
        fontSize: 16,
        lineHeight: 24,
        fontWeight: "600",
    },
    title: {
        fontSize: 32,
        fontWeight: "bold",
        lineHeight: 32,
    },
    subtitle: {
        fontSize: 20,
        fontWeight: "bold",
    },
    link: {
        lineHeight: 30,
        fontSize: 16,
    },
});
