// thirdweb/dist/esm/exports/extensions/modular.js
/**
 * READ
 */
export { getInstalledExtensions, isGetInstalledExtensionsSupported, } from "../../extensions/modular/__generated__/ModularCore/read/getInstalledExtensions.js";
export { getSupportedCallbackFunctions, isGetSupportedCallbackFunctionsSupported, } from "../../extensions/modular/__generated__/ModularCore/read/getSupportedCallbackFunctions.js";
export { hasAnyRole, } from "../../extensions/modular/__generated__/ModularCore/read/hasAnyRole.js";
export { hasAllRoles, } from "../../extensions/modular/__generated__/ModularCore/read/hasAllRoles.js";
export { owner } from "../../extensions/modular/__generated__/ModularCore/read/owner.js";
export { ownershipHandoverExpiresAt, } from "../../extensions/modular/__generated__/ModularCore/read/ownershipHandoverExpiresAt.js";
export { rolesOf, } from "../../extensions/modular/__generated__/ModularCore/read/rolesOf.js";
/**
 * Write
 */
export { cancelOwnershipHandover } from "../../extensions/modular/__generated__/ModularCore/write/cancelOwnershipHandover.js";
export { completeOwnershipHandover, } from "../../extensions/modular/__generated__/ModularCore/write/completeOwnershipHandover.js";
export { grantRoles, } from "../../extensions/modular/__generated__/ModularCore/write/grantRoles.js";
export { installExtension, isInstallExtensionSupported, } from "../../extensions/modular/__generated__/ModularCore/write/installExtension.js";
export { renounceOwnership } from "../../extensions/modular/__generated__/ModularCore/write/renounceOwnership.js";
export { requestOwnershipHandover } from "../../extensions/modular/__generated__/ModularCore/write/requestOwnershipHandover.js";
export { renounceRoles, } from "../../extensions/modular/__generated__/ModularCore/write/renounceRoles.js";
export { revokeRoles, } from "../../extensions/modular/__generated__/ModularCore/write/revokeRoles.js";
export { transferOwnership, } from "../../extensions/modular/__generated__/ModularCore/write/transferOwnership.js";
export { uninstallExtension, isUninstallExtensionSupported, } from "../../extensions/modular/__generated__/ModularCore/write/uninstallExtension.js";
export { getExtensionConfig } from "../../extensions/modular/__generated__/ModularExtension/read/getExtensionConfig.js";
export { installPublishedExtension, } from "../../extensions/modular/ModularCore/write/installPublishedExtension.js";
export { uninstallPublishedExtension, } from "../../extensions/modular/ModularCore/write/uninstallPublishedExtension.js";
export { uninstallExtensionByProxy, } from "../../extensions/modular/ModularCore/write/uninstallExtensionByProxy.js";
