// thirdweb/dist/cjs/extensions/modular/ModularCore/write/installPublishedExtension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.installPublishedExtension = void 0;
const bootstrap_js_1 = require("../../../../contract/deployment/utils/bootstrap.js");
const installExtension_js_1 = require("../../__generated__/ModularCore/write/installExtension.js");
function installPublishedExtension(options) {
    const { client, chain, account, contract, extensionName, constructorParams, publisherAddress, extensionData, } = options;
    return (0, installExtension_js_1.installExtension)({
        contract,
        asyncParams: async () => {
            const deployedExtension = await (0, bootstrap_js_1.getOrDeployInfraForPublishedContract)({
                chain,
                client,
                account,
                contractId: extensionName,
                constructorParams: constructorParams || [],
                publisher: publisherAddress,
            });
            return {
                extension: deployedExtension.implementationContract.address,
                data: extensionData || "0x",
            };
        },
    });
}
exports.installPublishedExtension = installPublishedExtension;
