// thirdweb/dist/cjs/extensions/modular/ModularCore/write/uninstallPublishedExtension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.uninstallPublishedExtension = void 0;
const infra_js_1 = require("../../../../contract/deployment/utils/infra.js");
const uninstallExtension_js_1 = require("../../__generated__/ModularCore/write/uninstallExtension.js");
function uninstallPublishedExtension(options) {
    const { client, chain, contract, extensionName, publisherAddress, constructorParams, extensionData, } = options;
    return (0, uninstallExtension_js_1.uninstallExtension)({
        contract,
        asyncParams: async () => {
            const deployedExtension = await (0, infra_js_1.getDeployedInfraContract)({
                chain,
                client,
                contractId: extensionName,
                constructorParams: constructorParams || [],
                publisher: publisherAddress,
            });
            return {
                extension: deployedExtension?.address,
                data: extensionData || "0x",
            };
        },
    });
}
exports.uninstallPublishedExtension = uninstallPublishedExtension;
