// thirdweb/dist/cjs/extensions/modular/__generated__/IModularCore/write/uninstallExtension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.uninstallExtension = exports.encodeUninstallExtension = exports.encodeUninstallExtensionParams = exports.isUninstallExtensionSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x42b7d0c8";
const FN_INPUTS = [
    {
        name: "extensionContract",
        type: "address",
        internalType: "address",
    },
    {
        name: "data",
        type: "bytes",
        internalType: "bytes",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `uninstallExtension` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `uninstallExtension` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isUninstallExtensionSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isUninstallExtensionSupported(contract);
 * ```
 */
async function isUninstallExtensionSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isUninstallExtensionSupported = isUninstallExtensionSupported;
/**
 * Encodes the parameters for the "uninstallExtension" function.
 * @param options - The options for the uninstallExtension function.
 * @returns The encoded ABI parameters.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeUninstallExtensionParams } "thirdweb/extensions/modular";
 * const result = encodeUninstallExtensionParams({
 *  extensionContract: ...,
 *  data: ...,
 * });
 * ```
 */
function encodeUninstallExtensionParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [
        options.extensionContract,
        options.data,
    ]);
}
exports.encodeUninstallExtensionParams = encodeUninstallExtensionParams;
/**
 * Encodes the "uninstallExtension" function into a Hex string with its parameters.
 * @param options - The options for the uninstallExtension function.
 * @returns The encoded hexadecimal string.
 * @extension MODULAR
 * @example
 * ```ts
 * import { encodeUninstallExtension } "thirdweb/extensions/modular";
 * const result = encodeUninstallExtension({
 *  extensionContract: ...,
 *  data: ...,
 * });
 * ```
 */
function encodeUninstallExtension(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeUninstallExtensionParams(options).slice(2));
}
exports.encodeUninstallExtension = encodeUninstallExtension;
/**
 * Prepares a transaction to call the "uninstallExtension" function on the contract.
 * @param options - The options for the "uninstallExtension" function.
 * @returns A prepared transaction object.
 * @extension MODULAR
 * @example
 * ```ts
 * import { uninstallExtension } from "thirdweb/extensions/modular";
 *
 * const transaction = uninstallExtension({
 *  contract,
 *  extensionContract: ...,
 *  data: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function uninstallExtension(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.extensionContract, resolvedOptions.data];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.uninstallExtension = uninstallExtension;
