// thirdweb/dist/esm/extensions/modular/__generated__/ModularExtension/read/getExtensionConfig.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x7c173ecc";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        name: "",
        type: "tuple",
        internalType: "struct IExtensionConfig.ExtensionConfig",
        components: [
            {
                name: "registerInstallationCallback",
                type: "bool",
                internalType: "bool",
            },
            {
                name: "requiredInterfaces",
                type: "bytes4[]",
                internalType: "bytes4[]",
            },
            {
                name: "supportedInterfaces",
                type: "bytes4[]",
                internalType: "bytes4[]",
            },
            {
                name: "callbackFunctions",
                type: "tuple[]",
                internalType: "struct IExtensionConfig.CallbackFunction[]",
                components: [
                    {
                        name: "selector",
                        type: "bytes4",
                        internalType: "bytes4",
                    },
                ],
            },
            {
                name: "fallbackFunctions",
                type: "tuple[]",
                internalType: "struct IExtensionConfig.FallbackFunction[]",
                components: [
                    {
                        name: "selector",
                        type: "bytes4",
                        internalType: "bytes4",
                    },
                    {
                        name: "permissionBits",
                        type: "uint256",
                        internalType: "uint256",
                    },
                ],
            },
        ],
    },
];
/**
 * Checks if the `getExtensionConfig` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getExtensionConfig` method is supported.
 * @extension MODULAR
 * @example
 * ```ts
 * import { isGetExtensionConfigSupported } from "thirdweb/extensions/modular";
 *
 * const supported = await isGetExtensionConfigSupported(contract);
 * ```
 */
export async function isGetExtensionConfigSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getExtensionConfig function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULAR
 * @example
 * ```ts
 * import { decodeGetExtensionConfigResult } from "thirdweb/extensions/modular";
 * const result = decodeGetExtensionConfigResult("...");
 * ```
 */
export function decodeGetExtensionConfigResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getExtensionConfig" function on the contract.
 * @param options - The options for the getExtensionConfig function.
 * @returns The parsed result of the function call.
 * @extension MODULAR
 * @example
 * ```ts
 * import { getExtensionConfig } from "thirdweb/extensions/modular";
 *
 * const result = await getExtensionConfig({
 *  contract,
 * });
 *
 * ```
 */
export async function getExtensionConfig(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
